### PR TITLE
Add support for SNAPSHOT dependencies layer

### PR DESCRIPTION
**Pull request by loosebazooka**
Partial solution to #403 

The behavior of this change is to separate out snapshot and non-snapshot dependencies as their own layers but still writes them both to `/app/libs`.

So `SourceFilesConfiguration` only add a getter on location of the snapshot dependency files, but adds no extra configuration paramter for the location to write those files (it's just shared with regular dependencies)

### CODE REVIEW

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/EntrypointBuilder.java [hunk 1]
```diff
@@ -37,6 +37,7 @@
     ImmutableList<String> classPaths =
         ImmutableList.of(
             sourceFilesConfiguration.getDependenciesPathOnImage() + "*",
+            sourceFilesConfiguration.getSnapshotDependenciesPathOnImage() + "*",
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/EntrypointBuilder.java]: Is there any harm to putting the snapshot dependencies in the same `/app/libs/` directory?

Comment (coollog) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/EntrypointBuilder.java]: We also might not want to have the snapshot layer if there are no snapshot dependencies, so having that directory added to the classpath may cause directory not found exceptions?

Comment (loosebazooka) [reply jib-core/src/main/java/com/google/cloud/tools/jib/builder/EntrypointBuilder.java]: Yeah, I don't have any reason reason for picking a separate directory. Let me change it.

**Comment by patflynn** [comment]
+1

File: jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildAndCacheApplicationLayerStep.java [hunk 2]
```diff
@@ -63,6 +63,17 @@
                               sourceFilesConfiguration.getDependenciesPathOnImage())
                           .build(),
                       cache))
+              .add(
```
Comment (coollog) [inline jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildAndCacheApplicationLayerStep.java]: We should probably not add this if there are no snapshot dependencies.

File: jib-core/src/test/java/com/google/cloud/tools/jib/builder/TestSourceFilesConfiguration.java [hunk 3]
```diff
@@ -70,12 +81,51 @@ public String getClassesPathOnImage() {
     return EXTRACTION_PATH + "classes/";
   }
 
-  /** Lists the files in the {@code resourcePath} resources directory. */
-  private ImmutableList<Path> getFilesList(String resourcePath)
-      throws URISyntaxException, IOException {
-    try (Stream<Path> fileStream =
-        Files.list(Paths.get(Resources.getResource(resourcePath).toURI()))) {
-      return fileStream.collect(ImmutableList.toImmutableList());
+  public static Builder builder() {
```
Comment (coollog) [inline jib-core/src/test/java/com/google/cloud/tools/jib/builder/TestSourceFilesConfiguration.java]: ordering nit: for convention in the codebase, we're trying to follow the class element ordering of:
- public->private
- static->non-static
- members->constructor->methods

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleSourceFilesConfiguration.java [hunk 4]
```diff
@@ -92,11 +94,16 @@ private GradleSourceFilesConfiguration(Project project, GradleBuildLogger gradle
       if (resourcesOutputDirectory.equals(dependencyFile.toPath())) {
         continue;
       }
-      dependenciesFiles.add(dependencyFile.toPath());
+      if (dependencyFile.getName().contains("SNAPSHOT")) {
```
[the remark this answers is not in the document]
Comment (briandealwis) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleSourceFilesConfiguration.java]: It does look like Gradle supports [explicitly indicating that a dependency may be changing](
https://stackoverflow.com/a/42058780/600339) and [provides an API](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ExternalModuleDependency.html#isChanging--) ([docs](https://docs.gradle.org/current/userguide/declaring_dependencies.html#sub:declaring_dependency_with_changing_version)).  Opened #694.

**Pull request merged**